src/main/scala/dsls/regex/RegularExpression.scala

```diff
@@ -11,6 +13,35 @@ package dsls.regex
 abstract class RegularExpression {
   /** returns true if the given string matches this regular expression */
   def matches(string: String) = RegexMatcher.matches(string, this)
+
+  def ||(that: RegularExpression) = Union(this, that)
+
+  def ~(that: RegularExpression) = Concat(this, that)
+
+  // Kleene star - zero or more regex repetitions. Parameter-less method
+  def <*> = Star(this)
+
+  // One or more repetitions
+  def <+> = this ~ (this <*>)
+
+  // Returns n repetitions of the preceding pattern
+  // each repetition is filled in a list, then concatenated to the overall regex
+  def apply(n: Int) = {
+    val e: RegularExpression = EPSILON
```

Comment:
Did you need to specify the type of `e` because Scala wouldn't apply `Concat` to `EPSILON`? It seems like Scala should be able to implicitly convert it to a `RegularExpression` since it's one implicit conversion. But I might be missing other implicit conversions here.